ddtrace/context.py

```diff
@@ -134,76 +132,37 @@ def _set_current_span(self, span):
+            # Assume the first span added to the context is the local root
+            if self._local_root_span is None:
+                self._local_root_span = span
             self._set_current_span(span)
-            self._trace.append(span)
             span._context = self
 
     def close_span(self, span):
-        # type: (Span) -> Tuple[Optional[List[Span]], Optional[bool]]
-        """
-        Mark a span as a finished, increasing the internal counter to prevent
-        cycles inside _trace list.
+        # type: (Span) -> None
+        """Updates the context after a span has finished.
+
+        The next active span becomes `span`'s parent.
         """
         with self._lock:
-            self._finished_spans += 1
-
-            # Safe-guard: prevent the last current span from being set to the parent
-            # of any span but the top-level span.
-            # The situation this avoids is when a parent closes before a child
-            # and the child is the last to close in the trace. When this happens
-            # the current_span would otherwise be set to the child's parent which
-            # has already closed. The context will be reset but the current_span
-            # will still point to that child's parent which would cause subsequent
-            # spans to be parented incorrectly.
-            if self._finished_spans != len(self._trace) or span == self._trace[0]:
+            if span == self._local_root_span:
+                if self.dd_origin is not None:
+                    span.meta[ORIGIN_KEY] = self.dd_origin
+                if self._sampling_priority is not None:
+                    span.metrics[SAMPLING_PRIORITY_KEY] = self._sampling_priority
+
+            # If a parent exists to the closing span and it is unfinished, then
+            # activate it next.
+            if span._parent and not span._parent.finished:
                 self._set_current_span(span._parent)
-
-            if self._finished_spans == len(self._trace):
-                # get the trace
-                trace = self._trace
-                sampled = self._is_sampled()
-                sampling_priority = self._sampling_priority
-                # attach the sampling priority to the context root span
-                if sampled and sampling_priority is not None and trace:
-                    trace[0].set_metric(SAMPLING_PRIORITY_KEY, sampling_priority)
-                origin = self.dd_origin
-                # attach the origin to the root span tag
-                if sampled and origin is not None and trace:
-                    trace[0].meta[ORIGIN_KEY] = str(origin)
-
-                # clean the current state
-                self._trace = []
-                self._finished_spans = 0
+            # Else the span that is closing is the local root span or an error
+            # has been made in the context management (parent closed before a
+            # child so clear out the context to prepare for the next trace.
```

Comment:
do we want to log something if a parent has finished before the child?

Reply:
Definitely!